llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl

```diff
@@ -1,10 +1,175 @@
+TdFilesInfo = provider(
+    "Holds tablegen files and the dependencies and include paths necessary to" +
+    " build them.",
+    fields = {
+        "transitive_sources": "td files transitively used by this rule.",
+        "transitive_includes": (
+            "include arguments to add to the final tablegen invocation." +
+            " Relative paths are interpreted as relative to the current" +
+            " label's package. Absolute paths are interpreted as relative to" +
+            " the current label's workspace root."
+        ),
+    },
+)
+
+# For now we allow anything that provides DefaultInfo to just forward its files.
+# In particular, this allows filegroups to be used. This is mostly to ease
+# transition. In the future, the TdFilesInfo provider will be required.
+# TODO(gcmn): Switch to enforcing TdFilesInfo provider.
+def _get_dep_transitive_srcs(dep):
+    """Extract TdFilesInfo.transitive_sources, falling back to DefaultInfo.files."""
+    if TdFilesInfo in dep:
+        return dep[TdFilesInfo].transitive_sources
+    return dep[DefaultInfo].files
+
+def _get_dep_transitive_includes(dep):
+    """Extract TdFilesInfo.transitive_includes, falling back to an empty depset()."""
+    if TdFilesInfo in dep:
+        return dep[TdFilesInfo].transitive_includes
+    return depset()
+
+def _get_transitive_srcs(srcs, deps):
+    """Obtain the source files for a target and its transitive dependencies.
+
+    Args:
+      srcs: a list of source files
+      deps: a list of targets that are direct dependencies
+    Returns:
+      a collection of the transitive sources
+    """
+    return depset(
+        direct = srcs,
+        transitive = [_get_dep_transitive_srcs(dep) for dep in deps],
+    )
+
+def _get_transitive_includes(includes, deps):
+    """Obtain the includes paths for a target and its transitive dependencies.
+
+    Args:
+      includes: a list of include paths
+      deps: a list of targets that are direct dependencies
+    Returns:
+      a collection of the transitive include paths
+    """
+    return depset(
+        direct = includes,
+        transitive = [_get_dep_transitive_includes(dep) for dep in deps],
+    )
+
+def _resolve_includes(ctx, includes):
+    """Resolves include paths to paths relative to the execution root.
+
+    Relative paths are interpreted as relative to the current label's package.
+    Absolute paths are interpreted as relative to the current label's workspace
+    root."""
+    package = ctx.label.package
+    workspace_root = ctx.label.workspace_root
+    workspace_root = workspace_root if workspace_root else "."
+    resolved_includes = []
+    for include in includes:
+        if not include.startswith("/"):
+            include = "/" + package + "/" + include
+        include = workspace_root + include
+        resolved_includes.append(include)
+        resolved_includes.append(ctx.genfiles_dir.path + "/" + include)
```

Comment:
Only FYI: there's also ctx.bin_dir, which unless a bazel flag is used, is the same as ctx.genfiles_dir. By default ctx.actions.declare_file puts the file in the bin_dir. Sadly, internally in Google they are not the same. Just so you know in case you are wondering why some file is in genfiles in bazel but it's not in blaze.

Reply:
Yes I encountered this when porting to Blaze, which is why I added `output_to_genfiles` to this rule so I get consistent behavior in both. Does that seem like the correct workaround. I think only things in genfiles can be used by later rules, right? So since this is creating header files and such to be included later I do want it to output to genfiles (although as you note, by default in Bazel there's no difference):

> output_to_genfiles
> default = False
If true, the files will be generated in the genfiles directory instead of the bin directory. Unless you need it for compatibility with existing rules (e.g. when generating header files for C++), do not set this flag.

https://docs.bazel.build/versions/master/skylark/lib/globals.html#rule.output_to_genfiles

Reply:
I'm not sure I follow "I think only things in genfiles can be used by later rules, right?"

Reply:
So maybe at the end the problem will be solved by adding 3 includes by default - `include`, `ctx.genfiles_dir.path + "/" + include`, and `ctx.binfiles_dir.path + "/" + include`.

Reply:
Alternatively, we could not remember include directories in the provider, but all the possible included files? And then map files to their parent directories when creating args. Am I making sense? I am rushing this reply on a Friday afternoon :)

Reply:
Ah, that documentation made me thing that maybe if you were generating a C++ header you needed to make sure it output to genfiles. Yeah perhaps we need to add all three.

> Alternatively, we could not remember include directories in the provider, but all the possible included files? And then map files to their parent directories when creating args.

Well we already have all possible included files (that's the deps), but we still need to know the include search paths because the way in which they're included is going to be relative to those, not just by their basename. That is, we need to add the `mlir/include` directory to the include paths so that we can reference `mlir/include/mlir/IR/OpBase.td` as `mlir/IR/OpBase.td`. This is basically the same as how the C++ compiler works (which is why it would be super helpful if there was a Starlark implementation of that I could reference).

Reply:
Ok yeah, there are a number of rules that expect generated outputs to be generated in genfiles (since that's where genrule puts them). I think it may still be good to make this rule more tolerant and have its inclusion paths prefixed with all three, but also I think it needs to output to genfiles.